src/Compilers/Core/Portable/SourceGeneration/Nodes/CombineNode.cs

```diff
@@ -41,7 +41,8 @@ public NodeStateTable<(TInput1, TInput2)> UpdateStateTable(DriverStateTable.Buil
                 return previousTable;
             }
 
-            var builder = graphState.CreateTableBuilder(previousTable, _name, _comparer);
+            var totalEntryItemCount = input1Table.GetTotalEntryItemCount();
+            var builder = graphState.CreateTableBuilder(previousTable, _name, _comparer, totalEntryItemCount);
```

Comment:
sorry, I don't yet have a strong grasp of this area. But was wondering if there's an analogous change to also make for calls to this method in 'BatchNode.cs' line 119 and 'SourceOutputNode.cs' line 51.

Reply:
BatchNode already does something similar. I'll look at sourcenode!

Reply:
ok, looking into this, BatchNode doesn't need this.  It's effectively making an `ImmutableArray<T>[]` (an array of an array).  The outer array it creates will effectively always be one element.  So we're always going to pool that array builder just fine.   

THe important thing for BatchNode is that hte *inner* `ImmutableArray<T>` is not creating waste.  The work for that was already done in https://github.com/dotnet/roslyn/pull/62169

Reply:
SourceOutputNode looks like it could benefit from this optimization.  However, i haven't seen that guy show up in traces (yet), so i haven't invested there.  I will keep a look out if it comes up!